refactor(core): Generalize binary data manager interface (no-changelog)

ivov opened the pull request:
Depends on: #7092 | Story: [PAY-768](https://linear.app/n8n/issue/PAY-768)

This PR: 
- Generalizes the `IBinaryDataManager` interface.
- Adjusts `Filesystem.ts` to satisfy the interface.
- Sets up an S3 client stub to be filled in in the next PR.
- Turns `BinaryDataManager` into an injectable service.
- Adjusts the config schema and adds new validators.

Note that the PR looks large but all the main changes are in `packages/core/src/binaryData`.

Out of scope:
- `BinaryDataManager` (now `BinaryDataService`) and `Filesystem.ts` (now `fs.client.ts`) were slightly refactored for maintainability, but fully overhauling them is **not** the focus of this PR, which is meant to clear the way for the S3 implementation. Future improvements for these two should include setting up a backwards-compatible dir structure that makes it easier to locate binary data files to delete, removing duplication, simplifying cloning methods, using integers for binary data size instead of `prettyBytes()`, writing tests for existing binary data logic, etc.

netroy committed:
refactor(core): Simplify executions and binary data pruning

ivov committed:
Merge branch 'master' into pay-771-implement-soft-deletion

ivov committed:
Merge branch 'master' into pay-771-implement-soft-deletion

ivov committed:
Remove test code from service

ivov committed:
Use time constants

ivov committed:
Improve naming

ivov committed:
Use native typeorm soft deletion

ivov committed:
Improve naming

ivov committed:
Make batch size class field

ivov committed:
Cleanup

ivov committed:
Remove unused method

ivov committed:
Cleanup

ivov committed:
Filter out soft-deleted executions

ivov committed:
Merge branch 'master' into pay-771-implement-soft-deletion

ivov committed:
Add tests

ivov committed:
Improve test

ivov committed:
Soft-delete in single pass

ivov committed:
Restore value

ivov committed:
Restore from debug value

ivov committed:
Add clarifying comments

ivov committed:
Update tests

ivov committed:
Add `typedi` to core

ivov committed:
Update package.json

ivov committed:
Turn `BinaryDataManager` into service

ivov committed:
Add `object` to binary data schema options

ivov committed:
Use service throughout core and cli

ivov committed:
Update tests

ivov committed:
Reduce diff

ivov committed:
Update default

ivov committed:
Add docline

ivov committed:
Remove redundant output types

ivov committed:
Remove unused type

ivov committed:
Rename to submanagers to client

ivov committed:
Rename `BinaryDataFileSystem` to `FileSystemClient`

ivov committed:
Clean up types

ivov committed:
Move `binaryToBuffer` to manager

ivov committed:
Rename `BinaryDataManager` to `BinaryDataService`

ivov committed:
Unify storage path naming

ivov committed:
Set binary data service in nodes-base test helper

ivov committed:
Rename manager in nodes-base

ivov committed:
More renamings

ivov committed:
Generalize interface

ivov committed:
Continue generalizing

ivov committed:
Fix test

ivov committed:
Better naming

ivov committed:
More renamings

ivov committed:
Minor cleanup

ivov committed:
More minor cleanup

ivov committed:
Add clarifying comment

ivov committed:
Speed up pruning if high volume

ivov committed:
Merge branch 'master' into pay-771-implement-soft-deletion

ivov committed:
Remove call from hook

ivov committed:
Merge parent branch

ivov committed:
Refix conflict

ivov committed:
Add formats to schema

ivov committed:
Add `ensureStringArray` util

ivov committed:
Adjust types based on parsing

ivov committed:
Add type guard

ivov committed:
Clean up types in `init()`

ivov committed:
Fix test

ivov committed:
Remove unneeded modifier

ivov committed:
Cleanup

ivov committed:
Fix node tests

ivov committed:
Remove unneeded line

ivov committed:
Rename constant

ivov committed:
Remove comment

ivov committed:
Fix tests

ivov committed:
Cleanup

ivov committed:
Rename error

ivov committed:
Improve error message

ivov committed:
Better error

ivov committed:
Better message

ivov committed:
Make execution ID non-nullable

ivov committed:
Merge master

ivov committed:
Readability improvements

ivov committed:
Adjust types, followup to 67e163f

ivov committed:
Merge parent branch

ivov committed:
Fix lint

ivov committed:
Fix lint

ivov committed:
Cleanup

ivov commented on packages/core/src/binaryData/types.ts:
```diff
@@ -0,0 +1,49 @@
+
+	export type Config = InMemoryConfig | FileSystemConfig;
+
+	export interface Manager {
```
Will address leftover TODOs in next PR implementing object storage.

ivov committed:
Comment out cache keys

ivov committed:
Revert "Comment out cache keys"

This reverts commit 96ea9eab6de78a9b8e494beacdcf413161373e17.

ivov committed:
Rename dir to `TempBinaryData`

ivov committed:
Rename back to `BinaryData`

netroy commented on packages/core/src/BinaryData/BinaryData.service.ts:
```diff
@@ -0,0 +1,217 @@
+		await this.getManager(this.mode).deleteManyByExecutionIds(executionIds);
+	}
+
+	async duplicateBinaryData(inputData: Array<INodeExecutionData[] | null>, executionId: string) {
```
We have a lot of code around duplicating binary data, that's only triggered via `IExecuteFunctions['executeWorkflow']`, and that method doesn't seem to be called from anywhere. Maybe this is dead-code and should be removed.

But if this is not dead code, I don't think we should be cloning this data in filesystem (or object store). Binary data is is never edited, and we can pass a readonly reference to binary data wherever we expect a copy.

ivov replied:
Certainly there is a lot more tech debt to address (small listing in the description), but if this bit is not blocking the S3 implementation, then let's tackle this later.

Implementation-wise, if we return a read-only reference instead of cloning the binary data, if we delete an execution we'll delete its binary data, which means that the non-deleted execution will be having a dangling pointer to deleted binary data.

ivov committed:
Merge master

ivov committed:
Clear timers on shutdown

ivov commented on packages/cli/src/config/schema.ts:
```diff
@@ -907,7 +900,7 @@ export const schema = {
 			env: 'N8N_DEFAULT_BINARY_DATA_MODE',
 			doc: 'Storage mode for binary data',
 		},
-		localStoragePath: {
```
Will do. I didn't know the config file syntax was similar to `schema.ts`. It wish it were just a list of env vars.

ivov replied:
Renamed back to `localStoragePath` at [04f4b83](https://github.com/n8n-io/n8n/pull/7164/commits/04f4b838840ea3e40c261a3d38d618d55ade3be8), since I'd rather not have to bother users for this.

ivov committed:
Set timers only on main instance

ivov committed:
Also for clearing timers

ivov committed:
Merge parent branch

ivov committed:
Rename back to `localStoragePath`

ivov committed:
Remove unused arg

ivov commented on packages/core/src/BinaryData/BinaryData.service.ts:
```diff
@@ -0,0 +1,217 @@
+
+	private managers: Record<string, BinaryData.Manager> = {};
+
+	async init(config: BinaryData.Config, mainManager = false) {
```
[03264f9](https://github.com/n8n-io/n8n/pull/7164/commits/03264f941d04a2ca98279d6c456a701a21b33088)

ivov committed:
Fix lint

ivov committed:
Add logging, refactor for readability

ivov committed:
Ensure hard-deletion select includes soft-deleted rows

ivov committed:
Switch `info` to `debug`

ivov committed:
Fix tests

ivov committed:
Remove redundant checks for `deletedAt` being `NULL`

The column `deletedAt` is marked `@DeleteDateColumn()` so all reads
from the repository automatically add a `WHERE` clause checking
that the column `IS NULL`.

ivov committed:
Fix lint

ivov committed:
Fix last test

ivov committed:
More missing loggers in tests

ivov committed:
Add logger to even more test

ivov committed:
Refactor logging for tests

ivov committed:
Merge parent branch

ivov committed:
Merge master, fix conflicts

ivov committed:
Fix misresolved conflict

ivov committed:
Remove excess check on init

ivov committed:
Remove `ObjectStore.manager.ts` stub

netroy commented on packages/cli/src/config/schema.ts:
```diff
@@ -900,9 +893,9 @@ export const schema = {
 
-	binaryDataManager: {
+	binaryDataService: {
 		availableModes: {
```
Renaming this is a breaking change, and should be documented in `BREAKING-CHANGES.md`

netroy replied:
if we are reverting, can we revert all of the renaming, and not just the config? either that, or we don't revert and list this as a breaking change.

ivov replied:
I think there is benefit in a temporary middle ground. Preexisting code uses "manager" for both binary data manager and its modes, so it's rather confusing and inconsistent. With this division into service and managers, the distinction is clearer, and the inconsistency is reduced to only this handful of entrypoint config spots where the renaming would cause a breaking change. So on balance the bulk of code becomes clearer.

ivov committed:
Revert rename to `BinaryDataManager`

ivov committed:
Missing renaming in test

krynble reviewed (dismissed):
LGTM

ivov commented on packages/core/src/BinaryData/BinaryData.service.ts:
```diff
@@ -0,0 +1,217 @@
+		if (this.availableModes.includes('filesystem') && config.mode === 'filesystem') {
+			this.managers.filesystem = new FileSystemManager(config.localStoragePath);
+
+			await this.managers.filesystem.init();
```
Thank you, please see [20d5ea7](https://github.com/n8n-io/n8n/pull/7164/commits/20d5ea78494d75f4283eeba7c8f7d40f5a3465f3)

ivov committed:
Rename constant

ivov commented on packages/core/src/BinaryData/FileSystem.manager.ts:
```diff
@@ -0,0 +1,138 @@
+import type { BinaryData } from './types';
+
+const EXECUTION_ID_EXTRACTOR =
+	/^(\w+)(?:[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12})$/;
```
Love it, I didn't know this. Since there are no tests and we will be refactoring this in future, for now I'll keep it as it was beforehand.

ivov commented on packages/core/src/BinaryData/utils.ts:
```diff
@@ -0,0 +1,25 @@
+ * - `filesystem` (on disk)
+ * - `object` (S3)
+ */
+export const BINARY_DATA_MODES = ['default', 'filesystem', 'object'] as const;
```
Please see 0cdfb7901f49bcc4cec7ae428885b6fafaef6de0

krynble commented:
Small note: for some reason I'm getting this on my terminal:

```
Press "o" to open in Browser.
InvalidBinaryDataManagerError: No binary data manager found
InvalidBinaryDataManagerError: No binary data manager found
InvalidBinaryDataManagerError: No binary data manager found
```

When viewing executions. Any ideas @ivov ?

ivov committed:
Merge master

ivov committed:
Restore `LogCatch`

`LogCatch` was merged into master at a different file: `/packages/core/binaryData/index.ts`, which is now at `BinaryData.service.ts`

ivov committed:
Add file not found error to `getSize`

ivov commented:
> Small note: for some reason I'm getting this on my terminal:
> 
> ```
> Press "o" to open in Browser.
> InvalidBinaryDataManagerError: No binary data manager found
> InvalidBinaryDataManagerError: No binary data manager found
> InvalidBinaryDataManagerError: No binary data manager found
> ```
> 
> When viewing executions. Any ideas @ivov ?

Unable to reproduce, messaging you for more details.

ivov committed:
Add `mode` to `InvalidBinaryDataManagerError`

ivov committed:
Delete many by execution IDs only if manager available

krynble reviewed (approved):
LGTM

janober commented:
Got released with [`n8n@1.9.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.9.0)
